File: prefix-dl.pl
#!/usr/bin/perl -w

use strict;
use warnings;
use QtCore4;
use QtGui4;
use QtCore4::debug qw(signals); #qw(ambiguous);
use Config::Simple;
use Data::Dumper;
use URI::URL;


my $config_file = $ENV{'HOME'} . '/.prefixes';

$ENV{'TMP'} = defined($ENV{'TMP'}) ? $ENV{'TMP'} : $ENV{'HOME'} . '/.pdl-tmp';

mkdir $ENV{'TMP'};

#print $config_file;

my $config = load_config();

main_window();


sub load_config {


	my $config = new Config::Simple(syntax=>'simple');
	if( -e $config_file) { 
		$config->read($config_file); 
	} else {
		$config->{_FILE_NAME} = $config_file;
	}
	$config->autosave(1);

	if(!defined($config->param('prefixes'))) {
		my @default_prefixes = ('ana', 'lol', 'les', 'dem', 'elf');
#		my @default_prefixes = ('a', 'b', 'c', 'd', 'e', 'f', 'g');
		$config->param('prefixes', \@default_prefixes);
	}

	return $config;

}


sub main_window {
	my @args = ($0 . ' ' . $ENV{'PWD'});
	my $app = Qt::Application( \@args );
	my $bs = Qt::Frame();
	my $split = Qt::VBoxLayout($bs);

	$bs->setWindowTitle($ENV{'PWD'} . ' - prefix-dl.pl' );

	my $history = Qt::TextEdit();
	$history->setReadOnly(1);
	$split->addWidget($history);

	my $ia_frame = Qt::Frame();
	my $input_area = Qt::HBoxLayout($ia_frame);
	$split->addWidget($ia_frame);


	#my $hello = Qt::Label( 'Hello, World!' );
	#$hello->show();
	#my $button = Qt::PushButton( 'Go');
	#my $tb = Qt::LineEdit('');

	my $pref_chooser = new prefix_chooser_widget();

	$split->addWidget($pref_chooser);

	#update_prefix_list($split, 1);

	my $line = new worker_line($pref_chooser, $history);

	$split->addWidget($line);


	#$button_area

	$bs->show();

	#$pref_chooser->{'bmenu'}->menu->actions->[0]->trigger();

	#$container->show();

	exit $app->exec();

}

sub hbox {

	my $hb = Qt::Frame();
	my $box = Qt::HBoxLayout($hb);

	return $hb;

}

package worker_line {

	use QtCore4;
	use QtCore4::isa qw( Qt::LineEdit );
	use QtCore4::slots
		onreturn => [],
		txtchanged => ['QString'];
	use File::Basename;	
	use Cwd;
	use Data::Validate::URI qw(is_uri);
	use File::Fetch;
	use File::Copy;

	sub NEW {

		my $t = shift;
		my @args = @_;
		my $logging_area = pop @args;
		my $prefchooser = pop @args;
		$t->SUPER::NEW(@args);
		this->{'prefix_chooser'} = $prefchooser;
		this->{'logging_area'}   = $logging_area;
		my $cr = this->connect(this, SIGNAL('returnPressed()'), this,  SLOT('onreturn()'));
		#print "Connect: $cr\n";
		$cr = this->connect(this, SIGNAL('textChanged(QString)'), this, SLOT('txtchanged(QString)'));
		#print "Connect: $cr\n";

	}

	sub filename_part {
                my $url = shift;
                if ($url !~ /\/([^\/]+\.[^\/]+)$/) {
                        return undef;
                }
                my $trailer = $1;
                if ($trailer =~ /^(.*)\?.*$/) {
                        return $1;
                }
                return $trailer;
        }

	sub construct_target_filename {
		my $prefix = shift;
		my $url = shift;
		my $base_dir = shift;
		my $fname_part = filename_part($url);
		if (length($fname_part) > 255) {
			my @bits = split(/\./, $fname_part);
			my $right = pop @bits;
			my $left = join('.', @bits);
			$left = substr($left, 0, 255 - 2 - length($right) - length ($prefix));
			$fname_part = $left . '.' . $right;
		}
		return $base_dir . '/' . $prefix . '-' . $fname_part;
	}

	sub target_exists {
		my $prefix = shift;
		my $url = shift;
		my $base_dir = shift;
		return -e construct_target_filename($prefix, $url, $base_dir);
	}

	sub gen_referer {
		my $url = shift;
		my $url_obj = new URI::URL $url;
		return $url_obj->scheme . '://' . $url_obj->netloc;
	}

	sub onreturn {
		#print this->{'prefix_chooser'}->prefix() .  this->text() . "\n";
		if(!is_uri(this->text())) {
			this->{'logging_area'}->append("The entered text doesn't look like a URL.\n");
			return;
		}
		if(this->prefix() eq "") {
			this->{'logging_area'}->append("No prefix selected.\n");
			return;
		}

		my $cwd = getcwd();
		my $prefix = this->prefix();
		this->{'prefix_chooser'}->promote_prefix($prefix);
		my $url = this->text();
		my $target = construct_target_filename($prefix, $url, $cwd);
		if (target_exists($prefix, $url, $cwd)) {
			this->{'logging_area'}->append("-e: " . $target . "\n");
			this->clear();
			return;		
		}
		my $referer = gen_referer($url);

		my $cline = "wget -O '$target' --referer='$referer' '$url' 2>&1";
		#print $cline . "\n";
		my $wget_output = `$cline`;
		my $rcode = $?;
		if ($rcode != 0) {
			this->{'logging_area'}->append("\n\n" . $wget_output . "\n\n");
			if (target_exists($prefix, $url, $cwd)) {
				unlink $target;
			}
		}
		
		if(-e $target) {
			this->{'logging_area'}->append(this->prefix() . ': ' . $target . "\n");
			this->clear();
		}

	}

	sub txtchanged {
		my $txt = shift;
		#print "Txt: " . this->presumedOutputFilename()  .  "\n";
	}

	sub prefix {
		return this->{'prefix_chooser'}->prefix();
	}

	sub presumedOutputFilename {

		my $of = this->{'prefix_chooser'}->prefix() . '-';
		my $bn = fileparse(this->text());
		
		return $of . $bn;
	}

1;
};

package toggle_excl_button {

	use QtCore4;
	use QtCore4::isa qw( Qt::PushButton );
	use QtCore4::slots
		checkr => ['bool'];

	sub NEW {


		shift->SUPER::NEW(@_);# Qt::PushButton($txt);
		this->setCheckable(1);
		this->setAutoExclusive(1);
		this->connect(this, SIGNAL('clicked(bool)'), this, SLOT('checkr(bool)'));

	}

	sub checkr {

		my $checked = shift;

		#print this->text() . ' ' . this->isChecked() . "\n";

	}

1;
};

sub in_array {

	my $arr = shift;
	my $val = shift;

	my @array = @{ $arr };

	for(my $i = 0; $i <= $#array; ++$i) {
		if($array[$i] eq $val) {
			return 1==1;
		}
	}

	return 0==1;

}

sub move_prefix_to_front {

	my $pref = shift;

	my @prefixes = @{ $config->param('prefixes') };

	#print "Old prefixes:\n";
	#print Dumper(\@prefixes);

	my $found = (1==0);

	for(my $i = 0; $i <= $#prefixes; ++$i) {

		if($pref ne $prefixes[$i]) {
			next;
		}

		splice(@prefixes, $i, 1);
		$found = (1==1);
		last;
	}

	if(!$found) {
		return;
	}

	unshift @prefixes, $pref;
	$config->param('prefixes', \@prefixes);

	#print "New prefixes:\n";
	#print Dumper(\@prefixes);

}


package menu_item {

	use QtCore4;
	use QtCore4::isa qw( Qt::Action );
	use QtCore4::slots
		hoverin => [''],
		menu_item_clicked => [''];
	use QtCore4::debug qw(slots signals);

	sub NEW {

		my @args = @_;
		my $t = shift @args;
		if($#args < 0) {
			$args[0] = '';
		}
		if($#args < 1) {
			$args[1] = undef;
		}
		my $func = sub { };
		if($#args > 1) {
			$func = pop @args;
		}

		$t->SUPER::NEW(@args);
		my $crv;
		$crv = this->connect(this, SIGNAL('triggered()'), this, SLOT('menu_item_clicked()'));
		#print "Connect 1: " . ($crv ? 'true' : 'false') . "\n";
		$crv = this->connect(this, SIGNAL('hovered()'), this, SLOT('hoverin()'));
		#print "Connect 2: " . ($crv ? 'true' : 'false') . "\n";
		this->{'name'} = $args[0];
		this->{'update_function'} = $func;

		#print "Blocked signals: " . (this->signalsBlocked() ? 'true' : 'false') . "\n";
		#this->trigger();
		#hoverin();
		this;

	}

	sub hoverin {

		#print "Hovering " . this->{'name'} . "\n";

	}

	sub setName {

		my $name = shift;

		this->{'name'} = $name;	
	}

	sub menu_item_clicked {

		#print "Hi\n";

		my $func = this->{'update_function'};
		&$func(this);

	}
1;
}; #End Package

package menu_sublet {

	use QtCore4;
	use QtCore4::isa qw( Qt::Menu );
	use QtCore4::slots
		action_intermediary => ['QAction'];

	sub NEW {
		shift->SUPER::NEW(@_);

		this->connect(this, SIGNAL('triggered(QAction)'), this, SLOT('action_intermediary(QAction)'));
		
	}

	sub action_intermediary {

		my $action = shift;

		#print "Got action: $action\n";
		$action->trigger();

	}

}


package prefix_chooser_widget {

	use QtCore4;
	use QtCore4::isa qw( Qt::Frame );
	use Data::Dumper;

	sub NEW {

		shift->SUPER::NEW(@_);

		this->{'button_count'} = 7;
		this->{'buttons_per_row'} = 4;
		my @empty;
		this->{'buttons'} = \@empty;
		this->{'bmenu'}  = undef;
		#this->setLayout(Qt::HBoxLayout(this));
		this->setLayout(Qt::GridLayout(this));
		this->update_buttons();
		my $i = 0;
		for(; $i < this->{'button_count'}; ++$i) {
			my($row, $column) = row_column(this->{'buttons_per_row'}, $i);
			this->layout->addWidget(this->{'buttons'}->[$i], $row, $column, 1, 1);
		}
		my($row, $column) = row_column(this->{'buttons_per_row'}, $i);		
		this->layout->addWidget(this->{'bmenu'}, $row, $column, 1, 1);
	}

	sub row_column {
		my $columns = shift;
		my $val = shift;
		return int($val / $columns), $val % $columns;
	}


	sub prefix {

		my @b = @{ this->{'buttons'} };

		for(my $i = 0; $i <= $#b; ++$i) {
			if($b[$i]->isChecked()) {
				return $b[$i]->text();
			}
		}

		return "";

	}

	sub setCheckedButton {

		my $but = shift;
		my @b = @{ this->{'buttons'} };
		for(my $i = 0; $i <= $#b; ++$i) {

			$b[$i]->setChecked($i == $but);

		}

	}

	sub promote_prefix {
		my $prefix = shift;
		main::move_prefix_to_front($prefix);
                this->setCheckedButton(0);
                this->update_buttons();
	}

	sub update_buttons {

		my @prefixes = $config->param('prefixes');

		for(my $i = 0; $i < this->{'button_count'}; ++$i) {
	
			if(!defined(this->{'buttons'}->[$i])) {
				this->{'buttons'}->[$i] = new toggle_excl_button($prefixes[$i]);
			} else {
				this->{'buttons'}->[$i]->setText($prefixes[$i]);
			}
		
		}


		if(!defined(this->{'bmenu'})) {

			this->{'bmenu'} = Qt::PushButton("Others");
		}

		my $menu = new menu_sublet();

		my $selfless = this;

		for(my $i = this->{'button_count'}; $i <= $#prefixes; ++$i) {
                	$menu->addAction(new menu_item($prefixes[$i], $menu, sub {

                        	my $self = shift;
	                        my $nm = $self->{'name'};
        	                #print "Moving to front: " . $nm  . "\n";
                	        $selfless->promote_prefix($nm);
        	        }));
	        }

		$menu->addAction(new menu_item("New prefix", $menu, sub {

        	        my $ntxt = Qt::InputDialog::getText($menu, "New Prefix", "");
                	if(!defined($ntxt) || $ntxt eq "") {
	                        return;
        	        }

	                my @prefs = @{ $config->param('prefixes') };
        	        if(main::in_array(\@prefs, $ntxt)) {
                	        return;
	                }
        	        unshift @prefs, $ntxt;
                	$config->param('prefixes', \@prefs);
			$selfless->update_buttons();

        	}));
		this->{'bmenu'}->setMenu($menu);

		my $actions = this->{'bmenu'}->menu->actions();

        	#print Dumper($actions);
		foreach my $act (@{$actions}) {
			#print Dumper($act->isEnabled()) . "\n";
		}
		#$actions->[0]->trigger();
		#$actions->[0]->trigger();
		#$actions->[0]->trigger();
		#print Dumper(this->{'bmenu'}->menu());
		#print "blah\n";

	}

1;
};

File: prefix-dl.py
#!/usr/bin/env python3

import sys
import os
import re
import json
import subprocess
import urllib

from django.core.validators import URLValidator
from django.core.exceptions import ValidationError

from PyQt5.QtWidgets import (QApplication, QLabel, QPushButton,
                             QVBoxLayout, QWidget, QTextEdit,
		             QGridLayout, QMenu, QAction, QInputDialog,
			     QLineEdit)

def is_url(uri):
  try:
    URLValidator()(uri)
    return True
  except:
    return False


class ConfigController(object):

  def __init__(self, filename):
    self._filename = filename
    # Stored with most-recent last
    self._prefixes = []
    self._known_prefixes = set(self._prefixes)
    self._filename = filename
    self._try_load()

  def add_prefix(self, prefix):
    if prefix in self._known_prefixes:
      return
    self._prefixes.append(prefix)
    self._known_prefixes.add(prefix)
    self._save()

  def to_front(self, prefix):
    if prefix not in self._known_prefixes:
      return
    self._prefixes.remove(prefix)
    self._prefixes.append(prefix)
    print("Promoting prefix %s" % prefix, "New ordering is [%s]" % self._prefixes)

  def get_prefix(self, index):
    if index >= len(self._prefixes):
      return None
    return self._prefixes[len(self._prefixes) - 1 - index]

  def _try_load(self):
    filename = self._filename
    if not os.path.exists(filename):
      return
    try:
      with open(filename) as f:
        dt = json.load(f)
        if self._looks_valid(dt):
          self._prefixes = dt
          self._known_prefixes = set(dt)
    except Exception as e:
      print("Unable to open prefix file?\n\n%s" % e)

  def _looks_valid(self, parsed_data):
    return ((type(parsed_data) is list) and
	    all(map(lambda x: type(x) is str, parsed_data)))

  def _save(self):
    with open(self._filename, mode='w') as f:
      json.dump(self._prefixes, f)

  def prefix_count(self):
    return len(self._prefixes)


def row_column(index, per_row):
  column, row = index % per_row, index // per_row
  return row, column


class PrefixChooser(QWidget):

  def __init__(self, controller, button_count=7, buttons_per_row=4):
    super(PrefixChooser, self).__init__()
    self._controller = controller
    self._button_count = button_count
    self._buttons_per_row = buttons_per_row
    self.refresh()
    self._selected_prefix = None

  def get_selected(self):
    return self._selected_prefix

  def refresh(self):
    self._buttons = []
    self._other_menu = None
    layout = QGridLayout()
    for i in range(self._button_count):
      row, column = row_column(i, self._buttons_per_row)
      text = self._controller.get_prefix(i)
      self._buttons.append(self._make_button(text))
      layout.addWidget(self._buttons[i], row, column, 1, 1)

    obutton_row, obutton_column = row_column(self._button_count,
					     self._buttons_per_row)

    layout.addWidget(self._make_other_button(),
	             obutton_row, obutton_column, 1, 1)
    if self.layout():
      QWidget().setLayout(self.layout())
    self.setLayout(layout)

  def _make_button(self, text):
    button = QPushButton(text)
    button.setCheckable(True)

    def on_click():
      print("Clicked %s" % text)
      for other in self._buttons:
        other.setChecked(False)

      button.setChecked(True)
      self._selected_prefix = text

    button.clicked.connect(on_click)
    return button

  def _make_other_button(self):
    button = QPushButton("Others")
    button.setMenu(OtherPrefixMenu(self._controller, self._button_count, self))
    return button

  def select_first(self):
    self._buttons[0].click()
    print("Newly selected: %s" % self._selected_prefix)


class OtherPrefixMenu(QMenu):

  def __init__(self, controller, button_count, parent):
    super(OtherPrefixMenu, self).__init__()
    self._controller = controller
    self._button_count = button_count
    self._parent = parent
    self.refresh()

  def refresh(self):
    for i in range(self._button_count, self._controller.prefix_count()):
      prefix = self._controller.get_prefix(i)
      self.addAction(self._make_menu_item(prefix))

    self.addAction(self._make_new_prefix_item())

  def _make_menu_item(self, text):
    action = QAction(text, self)

    def clicked():
      self._controller.to_front(text)
      self._parent.refresh()
      self._parent.select_first()

    action.triggered.connect(clicked)
    return action

  def _make_new_prefix_item(self):
    action = QAction("New Prefix", self)

    def clicked():
      value, ok = QInputDialog.getText(self, "New Prefix", "")

      if not value:
        return

      self._controller.add_prefix(value)
      self._parent.refresh()
      self._parent.select_first()

    action.triggered.connect(clicked)
    return action


def filename_part_of_url(url):
  m = re.search("\/([^\/]+\.[^\/]+)$", url)
  if not m:
    return None
  end = m.group(1)
  m = re.search("^(.*)\?.*$", end)
  if m:
    return m.group(1)
  return end


def construct_target_filename(prefix, url, outdir):
  filename_part = filename_part_of_url(url)
  if len(filename_part) > 200:
    bits = filename.split(".")
    end = bits[-1]
    filename_part = "".join(bits[:-1]).substr(0, 200) + "." + end
  return "%s/%s-%s" % (outdir, prefix, filename_part)


def generate_referer(url):
  parts = urllib.parse.urlparse(url)
  return parts.scheme + "://" + parts.netloc


def download_file(url, outfile):
  referer = generate_referer(url)
  result = subprocess.run(["wget",
                           "-O", outfile,
                           "--referer=%s" % referer,
                           url],
                          capture_output=True)

  if result.returncode == 0:
    return (True, None)
  else:
    return (False, "%s\n\n%s" % (result.stdout, result.stderr))


class WorkerLine(QLineEdit):

  def __init__(self, get_prefix_func, write_log_func, base_dir):
    self._get_prefix = get_prefix_func
    self._write_log = write_log_func
    self._base_dir = base_dir
    super(WorkerLine, self).__init__()
    self._setup()

  def _setup(self):
    self.returnPressed.connect(self._on_return)

  def _on_return(self):
    url = self.text()
    if not self._get_prefix():
      self._write_log("No prefix selected.")
      return
    if not is_url(url):
      self._write_log("The entered text doesn't look like a URL.")
      return
    outfile = construct_target_filename(self._get_prefix(), url, self._base_dir)

    if os.path.exists(outfile):
      self._write_log("-e: %s" % outfile);
      self.clear()
      return
    result, message = download_file(url, outfile)
    if result:
      self._write_log("%s: %s" % (self._get_prefix(), outfile))
      self.clear()
    else:
      self._write_log(message)



class HistoryArea(QTextEdit):

  def __init__(self):
    super(HistoryArea, self).__init__()
    self.setReadOnly(True)

  def WriteHistoryEntry(self, text):
    self.append(text)
    self.append("\n")


class DownloadWindow(QWidget):

  def __init__(self, controller, save_path):
    super(DownloadWindow, self).__init__()
    self._controller = controller
    self._save_path = save_path
    self._setup_widgets()

  def _setup_widgets(self):
    outermost_layout = QVBoxLayout()
    self._history = HistoryArea()
    outermost_layout.addWidget(self._history)
    self.setLayout(outermost_layout)
    self._chooser = PrefixChooser(self._controller)
    outermost_layout.addWidget(self._chooser)
    worker_line = WorkerLine(lambda: self._chooser.get_selected(),
			     self._history.WriteHistoryEntry,
			     self._save_path)
    outermost_layout.addWidget(worker_line)



if __name__ == '__main__':
  app = QApplication([])
  prefix_file = os.environ['HOME'] + "/.prefix_dl_py_prefixes"
  controller = ConfigController(prefix_file)
  widget = DownloadWindow(controller, os.getcwd())
  widget.resize(400, 600)
  widget.show()
  sys.exit(app.exec_())
